core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -280,38 +280,48 @@ private[kafka] object LogValidator extends Logging {
 
     var uncompressedSizeInBytes = 0
 
-    val batch = validateOneBatchRecords(records)
+    // Assume there's only one batch with compressed memory records; otherwise, return InvalidRecordException
+    // One exception though is that with format smaller than v2, if sourceCodec is noCompression, then each batch is actually
+    // a single record so we'd need to special handle it by creating a single wrapper batch that includes all the records
+    val batches: Iterable[MutableRecordBatch] = if (sourceCodec == NoCompressionCodec && records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V1)) {
```

Comment:
Potentially we could just check the magic of the first batch.

Reply:
I actually prefer to do this because:

1) for compressed records, we will only have one batch so we do not execute any unnecessary CPUs.

2) for uncompressed records and v1/v0, it's better to make sure they are all having the same magic byte anyways.